packages/core/src/sso/SamlConnector/utils.ts

```diff
@@ -0,0 +1,168 @@
+import { XMLValidator } from 'fast-xml-parser';
+import { got } from 'got';
+import * as saml from 'samlify';
+import { z } from 'zod';
+
+import {
+  samlMetadataGuard,
+  type SamlMetadata,
+  type BaseSamlConnectorConfig,
+  type ProfileMap,
+  MetadataType,
+  type BaseSamlConfig,
+} from '../types/saml.js';
+
+type ESamlHttpRequest = Parameters<saml.ServiceProviderInstance['parseLoginResponse']>[2];
+
+const xmlValidator = (xml: string) => {
+  try {
+    XMLValidator.validate(xml, {
+      allowBooleanAttributes: true,
+    });
+  } catch (error: unknown) {
+    throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, error);
+  }
+};
+
+const parseXmlMetadata = (xml: string): SamlMetadata => {
+  xmlValidator(xml);
+
+  // eslint-disable-next-line new-cap
+  const idP = saml.IdentityProvider({ metadata: xml });
+
+  const rawSingleSignOnService = idP.entityMeta.getSingleSignOnService(
+    saml.Constants.namespace.binding.redirect
+  );
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+  const singleSignOnService =
+    typeof rawSingleSignOnService === 'string'
+      ? rawSingleSignOnService
+      : Object.entries(rawSingleSignOnService).find(
+          ([key, value]) => key === saml.Constants.namespace.binding.redirect
+        )?.[1];
+  const rawSamlMetadata = {
+    entityId: idP.entityMeta.getEntityID(),
+    /**
+     * See implementation in `samlify` {@link https://github.com/tngan/samlify/blob/55f845da60b18d40668885c7f7e71ed0967ef67f/src/entity.ts#L88}.
+     */
+    nameIdFormat: idP.entitySetting.nameIDFormat,
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    signInEndpoint: singleSignOnService,
+    signingAlgorithm: idP.entitySetting.requestSignatureAlgorithm,
+    // The type inference of the return type of `getX509Certificate` is any, will be guarded by later zod parser if it is not string-typed.
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    x509Certificate: idP.entityMeta.getX509Certificate(saml.Constants.wording.certUse.signing),
+  };
+
+  // The return type of `samlify`
+  const result = samlMetadataGuard.safeParse(rawSamlMetadata);
+
+  if (!result.success) {
+    throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, result.error);
+  }
+
+  return result.data;
+};
+
+export const getRawSamlConfig = async (config: BaseSamlConnectorConfig): Promise<string> => {
+  if (config.metadataType === MetadataType.URL) {
+    const { body } = await got.get(config.metadataUrl);
+
+    const result = z.string().safeParse(body);
+
+    if (!result.success) {
+      throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, result.error);
+    }
+
+    return result.data;
+  }
+
+  return config.metadataXml;
+};
+
+export const fetchSamlConfig = async (config: BaseSamlConnectorConfig) => {
+  const rawMetadata = await getRawSamlConfig(config);
+  return parseXmlMetadata(rawMetadata);
+};
+
+export const getUserInfoFromRawUserProfile = (
```

Comment:
Same here, we might need to store the raw unknown user attributes in the `SsoIdentities`

Reply:
This method is intended to get the `SocialUserInfo` which can be used for Logto `users` table. Added another method `getParsedSamlAssertion` to get the raw content extracted from SAML assertion.

Reply:
SsoIdentities could be a `SocialUserInfo` extended data type. SsoIdentity data has two usages, 
1. store directly in the table as the raw reference of the IdP provider issued data
2. sync with Logto user profile. 

For the first case we should keep the original copy of the whatever passed from the SsoIdentity data
For the second case，we programmatically read and sync the properties. 
So a `SocialUserInfo` extended data type would work perfectly?